teps/0055-migrate-podlatency-to-taskprelaunchlatency.md

```diff
@@ -0,0 +1,121 @@
+---
+status: proposed
+title: Migrate podLatency to taskPrelaunchLatency
+creation-date: '2021-01-06'
+last-updated: '2021-03-05'
+authors:
+- '@yaoxiaoqi'
+---
+
+# TEP-0055: Migrate `podLatency` to `taskPrelaunchLatency`
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+- [Alternatives](#alternatives)
+  - [Keeping the metrics <code>podLatency</code>](#keeping-the-metrics-)
+  - [Using a map to store the pod <code>startRunning</code> time](#using-a-map-to-store-the-pod--time)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes to discard the metrics `podLatency` and add another metrics to
+report the pod creation -> pod running latency.
+
+## Motivation
+
+Currently, `podLatency` calculates the duration between pod creation time and
+pod scheduled time. The `podLatency` value is 0 at 99% of the time because
+Kubernetes timestamps are only at the resolution of a second, so if something
+takes less than a second it looks like it took 0 seconds. At the rest 1% of
+time, the metrics value is 1 or 2, which is not very worth to report. Therefore,
+this TEP proposes to discard the metrics `podLatency` and add another metrics to
+report the pod creation -> pod running latency. It can indicate how long a
+TaskRun pod takes to get ready for running.
+
+### Goals
+
+- Discard the current metrics `podLatency`
+- Add a metrics called `taskPrelaunchLatency` to indicate the pod creation ->
+  pod running latency
+
+### Non-Goals
+
+- Add multiple metrics to indicate different phases of TaskRun initialization,
+  like TaskRun creation time -> pod scheduled time latency, pod scheduled time
+  -> pod running time, etc
+
+## Proposal
+
+As `podLatency` is discarded, `taskPrelaunchLatency` is introduced to report the
+duration between pod creation time, and the time when pod phase turns to
+`Running`. However, Knative doesn't collect and report when a pod starts
```

Comment:
s/Knative/Tekton ?

The time when a `TaskRun` starts running is not written directly in the status, but we do have `startedAt` for each step, which comes from the corresponding `pod` container field. So I don't think we need to store any extra info in the `TaskRun` status.

The time the pod is scheduled is stored in the pod status, I don't think we need to replicate that in the `TaskRun` status to be able to emit the metric?

What might be challenging it choosing the right set of conditions to make sure that we report the metric only once.

Reply:
Oh I meant Knative. Because I wanted to get the time when the `Pod` status turns into `running`, which I believe is not recorded directly anywhere. But after read your comments I figured that all information we need is collected and your method is more reasonable.